readme: add section on package removal

ajeddeloh opened the pull request:
Document pitfalls and useful tricks for discovering where a package is
used.

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+# Removing packages
+
+Be careful when removing packages. The following section offers tips for preventing
+breakage, but they are by no means exhaustive. Gentoo's portage utilities wont take
```
`won't`

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+
+Be careful when removing packages. The following section offers tips for preventing
+breakage, but they are by no means exhaustive. Gentoo's portage utilities wont take
+into account some pretty core packages like catalyst or glibc, so take extra care
```
Is `package` the right way to describe the catalyst infrastructure?

ajeddeloh replied:
Probably not. It's not clear to me how exactly it's used. We have a catalyst package (`dev-util/catalyst`) but I can't find a single place we install it or use it.

bgilbert replied:
`scripts/build_library/catalyst.sh`

ajeddeloh replied:
After talking with @dm0-, jenkins installs this. I'm adding it to `sdk-extras` (https://github.com/coreos/coreos-overlay/pull/2867) so we can track it and its dependencies.

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+
+`equery d -a <package-name>` will tell you what packages depend on `package-name`. It will
+also generate a lot of false positives, since it considers all dependencies for all use flags,
+even ones we dont use, such as `test`.
```
`don't`

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+
+## emerge --emptytree
+`emerge --pretend --verbose --emptytree <package-name>` _should_ give a list of all the dependencies
+for a given package. After removing the ebuild and package, make sure it's removal doesn't break
```
`its`

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+the packages `board-packages` and `sdk-depends`. Remember to use the `emerge-<arch>-usr` commands
+to check `board-packages` and `emerge` to check `sdk-depends`.
+
+The ebuild must be removed and the package must be removed locally or the pacakge
```
`package`

bgilbert commented on README.md:
```diff
@@ -18,9 +20,46 @@ If you'd prefer to use a local copy of the portage tree, you can point
+to check `board-packages` and `emerge` to check `sdk-depends`.
+
+The ebuild must be removed and the package must be removed locally or the pacakge
+will be silently elided in package list. To see if there are any packages installed
```
`elided from the`

ajeddeloh commented:
Fixed typos and expanded contractions, made some of the language more precise.

bgilbert commented on README.md:
```diff
@@ -18,9 +22,49 @@ If you'd prefer to use a local copy of the portage tree, you can point
+to check `board-packages` and `emerge` to check `sdk-depends`.
+
+The ebuild must be removed from `portage-stable` and the package must be removed locally or else
+the package will be silently elided in the dependency list. To see if there are any packages
```
This appears to say that if I leave the ebuild in place and leave the package installed, it won't show up in the dep list, which doesn't sound right. Is the intent to say that the problem occurs if I remove the ebuild but leave the package installed?

ajeddeloh commented:
Updated:
 - Language on emptytree behaviour
 - Fixed `egencache` command and monospaced it
 - Added `@system` to the list of things to check

Added:
 - Note about emptytree and virtuals
 - Note about running tests on Jenkins
 - Note about `update_metadata` script
 - Note about checking `coreos-overlay` for things like use flags or accept_keywords.

bgilbert commented on README.md:
```diff
@@ -18,9 +22,64 @@ If you'd prefer to use a local copy of the portage tree, you can point
-    egencache --update portage-stable
-
-Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+## Testing changes
```
nit: blank line

bgilbert commented on README.md:
```diff
@@ -18,9 +22,64 @@ If you'd prefer to use a local copy of the portage tree, you can point
-
-Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+## Testing changes
+If you work at CoreOS, or have Jenkins running with our configuration, you can make pull requests with your changes and
```
The first part isn't necessary. `If you're running the [jenkins-os](https://github.com/coreos/jenkins-os) Jenkins configuration, ...`

ajeddeloh commented:
Added one more little bit when talking about emptytree to mention packages needed by the bootstrapping process.

bgilbert commented on README.md:
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
+
+`emerge --pretend --verbose --emptytree <package-name>` _should_ give a list of all the dependencies
+for a given package. Use this to test if `board-packages`, `sdk-depends`, and `@system` can still be
+emerged after removing an ebuild and pacakge.
```
`package`

bgilbert commented on README.md:
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
+`sdk-depends`. Use both when checking `@system`.
+
+Furthermore, the SDK bootstrapping process uses a list of packages defined the by SDK profile's packages.build
+file. Install dev-util/catalyst and run /usr/lib64/catalyst/targets/stage1/build.py to get a list of packages
```
Missing backticks.

bgilbert commented on README.md:
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
+file. Install dev-util/catalyst and run /usr/lib64/catalyst/targets/stage1/build.py to get a list of packages
+needed for the boostrapping process, then run `emerge --emptytree` on that list.
+
+The ebuild must be removed from `portage-stable` _and_ the package must be removed locally. If only the
```
`A package's ebuild` would better introduce the subject change.

bgilbert commented on README.md:
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
+When making changes affecting virtual packages (removing a provider, changing use flags that affect
+a provider or virtual, etc), `emerge --emptytree` will always prefer an already installed
+provider (unless it is masked or otherwise disabled), so unmerge the current provider before
+running `emerge --emptytree` to ensure the virtuals are resolved correctly. Look at the virtual packages's
```
`package's`

bgilbert commented on README.md:
```diff
@@ -18,9 +22,72 @@ If you'd prefer to use a local copy of the portage tree, you can point
+## Testing changes
 
-Upstream Gentoo sources: http://sources.gentoo.org/gentoo-x86/
+If you have Jenkins running with our configuration, you can make pull requests with your changes and
```
This is okay, but `our configuration` seems unnecessarily possessive. IMO it reads better to just name the config repo and link to it.
